chalk-solve/src/clauses/builtin_traits/discriminant_kind.rs

```diff
@@ -0,0 +1,69 @@
+use crate::clauses::ClauseBuilder;
+use crate::{Interner, RustIrDatabase, TraitRef, WellKnownTrait};
+use chalk_ir::{AliasTy, Normalize, ProjectionTy, Substitution, Ty, TyKind, TyVariableKind};
+
+pub fn add_discriminant_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    self_ty: Ty<I>,
+) {
+    let interner = db.interner();
+
+    let can_determine_discriminant = match self_ty.data(interner).kind {
+        TyKind::Adt(..)
+        | TyKind::Array(..)
+        | TyKind::Tuple(..)
+        | TyKind::Slice(..)
+        | TyKind::Raw(..)
+        | TyKind::Ref(..)
+        | TyKind::Scalar(_)
+        | TyKind::Str
+        | TyKind::Never
+        | TyKind::FnDef(..)
+        | TyKind::Generator(..)
+        | TyKind::Closure(..)
+        | TyKind::GeneratorWitness(..)
+        | TyKind::Foreign(_)
+        | TyKind::Dyn(_)
+        | TyKind::Function(..)
+        | TyKind::InferenceVar(_, TyVariableKind::Integer)
+        | TyKind::InferenceVar(_, TyVariableKind::Float) => true,
+        TyKind::OpaqueType(..)
+        | TyKind::Alias(_)
+        | TyKind::BoundVar(_)
+        | TyKind::Placeholder(_)
+        | TyKind::AssociatedType(..)
+        | TyKind::Error
+        | TyKind::InferenceVar(..) => false,
+    };
+
+    if !can_determine_discriminant {
+        return;
```

Comment:
I'm not sure if this is the right rule here. This will mean that we *can't* prove the `DiscriminantKind`, but I feel like we may want to return an "ambiguous" result instead.  It's not that there *is no* discriminant kind, it's that we don't know what it is.

Reply:
I guess the thing to do is to make it .. uh .. flounder for these sorts of self types.